### DISCUSSION TITLE
How do I use the debugger with Rust?

### DISCUSSION THREAD

**Discussion by David-Else** [helix-editor/helix#4048] [Q&A]
Hi, I have `codelldb` installed using the WIKI instructions https://github.com/helix-editor/helix/wiki/Debugger-Configurations#install-debuggers and want to debug a Rust program. I am quite new to debugging, I have done it in VS Code. How do I do it in Helix? I have added:

```
[[language]]
name = "rust"

[language.debugger]
command = "codelldb"
name = "codelldb"
port-arg = "--port {}"
transport = "tcp"

[[language.debugger.templates]]
name = "binary"
request = "launch"
[[language.debugger.templates.completion]]
completion = "filename"
name = "binary"

[language.debugger.templates.args]
program = "{0}"
runInTerminal = true
```

`launch debug binary` asks for binary, I enter the target and get:

![Screenshot from 2022-09-30 12-18-38](https://user-images.githubusercontent.com/12832280/193258662-df006bb4-bdcc-4729-8808-4f6fbcbb41c3.png)

I am on RHEL 9 Linux using Kitty, same problem with Debian. Is it still hard coded to use Tmux?

What about the default `lldb-vscode`? Why are there instructions to install `codelldb` but not `lldb-vscode`? How do I get a `lldb-vscode` binary, do I need to install VS Code?

The WIKI is incredibly confusing. Could someone please explain how to debug Rust? THANKS!

**Comment by kirawi** [comment]
`lldb-vscode` is distributed with LLVM as an executable. The wiki entry was created after the fact by someone else.

**Reply by ArikRahman** [reply]
Is the wiki reliable?

**Reply by David-Else** [reply]
@kirawi I have managed to download the binary for LLVM from https://github.com/llvm/llvm-project/releases/download/llvmorg-15.0.2/clang+llvm-15.0.2-x86_64-unknown-linux-gnu-rhel86.tar.xz and extracted the file `lldb-vscode` to `~/bin`. Now when I run I get:


![Screenshot from 2022-10-10 15-34-28](https://user-images.githubusercontent.com/12832280/194892950-0922eb7b-fa6c-46c7-a3b4-4dc4c55600a6.png)

I tried to copy the lib directory into `bin` too, no luck. Am I on the right track, how should I install the binary? Thanks!

**Reply by kirawi** [reply]
I'm not sure how it goes on Linux, but for me `liblldb` was in the `LLVM/bin` folder. Maybe `liblldb.so` has to be in the same folder as the executable?

**Reply by David-Else** [reply]
@kirawi I have that `lldb-vscode` fine and Helix sees it, the problem is the missing lib file `liblldb.so.15`, I can find it in the `lib` directory along with a symlink that seems to point to it.. it is all a bit ridiculous to get the debugger working.

Here is my bin dir, it won't find the lib:

![Screenshot from 2022-10-10 16-07-43](https://user-images.githubusercontent.com/12832280/194898060-45d3c5ec-bfe6-45be-88c1-a67664270f1a.png)

**Reply by kirawi** [reply]
I'm not sure. Here's the documentation for vscode: https://github.com/llvm/llvm-project/tree/main/lldb/tools/lldb-vscode#installation-for-visual-studio-code. Perhaps there's somewhere on LLVM you can ask.

**Comment by kirawi** [comment]
I checked, and you can build codelldb yourself: https://github.com/vadimcn/vscode-lldb/blob/fe1fb617b273d846b76ba75087e7f35232c9bd9e/BUILDING.md

**Reply by David-Else** [reply]
But I thought I wanted `lldb-vscode`, which is default for Helix, not `vscode-lldb`? The WIKI says:

>Note, the names can be confusing. vscode-lldb is a separate project from the aforementioned lldb-vscode

I read https://github.com/llvm/llvm-project/tree/main/lldb/tools/lldb-vscode#installation-for-visual-studio-code and it make no sense, I am trying to copy the right files, but they seem to be links to links in the downloaded binary, and then they don't load in Helix.
![Screenshot from 2022-10-10 23-41-56](https://user-images.githubusercontent.com/12832280/194962957-a86f24e6-12d3-477f-83e4-43a856ea86ca.png)

I assume at least one person, maybe @archseer has `lldb-vscode` working on Linux?

**Reply by the-mikedavis** [reply]
The lldb-vscode binary in Nix is defined here https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/compilers/llvm/14/lldb/default.nix (it's included in this repository's devShell in the flake, so it's probably what archseer is using). It looks like a somewhat normal CMAKE build so I assume you just need to follow the build instructions in the LLDB repo

**Reply by David-Else** [reply]
@the-mikedavis I have made some progress and added it to the WIKI:

https://github.com/helix-editor/helix/wiki/Debugger-Configurations#addendum-for-users-who-cant-make-any-of-the-above-work

I think the entire thing needs re-writing properly and then moving to the documentation section WITH how to actually use the debugger :)

**Reply by weiying-chen** [reply]
@David-Else Thanks for the instructions. But I still don't know how to launch the debugger.

I did `:debug-start`, and I got: `Failed to start debug session: cannot find binary path`.

I couldn't even select anything as instructed:

> Now when you run the debugger in Helix select launch debug target and binary, then for example, to debug Rust, target/debug/ and the name of your executable.

**Reply by David-Else** [reply]
@alexcheninfo I have given up on the debugger, it seems pre-beta and experimental at best :( You should be able to get it running following the instructions I made at https://github.com/helix-editor/helix/wiki/Debugger-Configurations#addendum-for-users-who-cant-make-any-of-the-above-work , at least on Linux. I have no other OS to test it on.

I have asked for help on so many occasions and nobody wants to talk about the debugger, I have no idea if anyone is even using it! I got it running, then to my horror the local variables were reported as memory addresses rather than values! Not sure how that is meant to help debugging?

I don't know why development stopped, it seems it took a year of work to merge it, and then... ?

**Reply by weiying-chen** [reply]
@David-Else 

I see, I'm about to give up too, ha. I'm curious, though, how did you run the debugger?

As I said, I did `:debug-start`, but I got: `Failed to start debug session: cannot find binary path.` I couldn't select anything as you instructed.

**Comment by David-Else** [comment]
Go to your terminal and type `lldb-vscode --help` to check you have it in your path.

You need to select the built binary for your application in the `target/debug/` directory. Auto-complete for the file path might not work due to your `.gitignore`, you can find it if you type `target/` and then you can see auto-complete to find `debug` and your binary.

**Reply by weiying-chen** [reply]
Sorry, I'm still confused.
1. Where is this `target/debug` directory?
2. By selecting, you mean selecting it from inside Helix?

By the way, `lldb-vscode` is recognize by Helix now:

    Binary for debug adapter: /home/alex/.local/bin/lldb-vscode

But when I open a `.rs` file and run `:debug-start` inside Helix, I get: Async job failed: none.

**Reply by David-Else** [reply]
> Where is this target/debug directory?

I am assuming you are debugging Rust? When you compile your app with `cargo build` the executable with debug code is created in `/target/debug/filename_of_your_project` in your project folder. This is the file you need to open to debug, it is different from languages like JavaScript where you select the source file, you need to select the actual compiled file in Rust.

> By selecting, you mean selecting it from inside Helix?

Yes.

**Reply by weiying-chen** [reply]
I already figured out how. With `Space` (Space Mode), `g` (Debug), then `l` (launch). This let's you select a Rust binary file with autocomplete. Thanks for helping and replying so fast!

By the way, the debugger works fine. The breakpoints work fine. But as you said,  the local variables are reported as memory addresses rather than values. I wonder if this is a feature or a bug:

https://i.stack.imgur.com/o7eWO.png

**Reply by rkshthrmsh** [reply]
Hi @weiying-chen and @David-Else, I was wondering if you were somehow able to access the local variable values from the debugger?

**Reply by David-Else** [reply]
@rkshthrmsh  No I was not, so I added it as an official feature request https://github.com/helix-editor/helix/issues/7007. It is very strange to me that it is not something everyone is talking about, maybe nobody uses a debugger?

**Comment by weiying-chen** [comment]
@rkshthrmsh No, I wasn't able. They are still reported as memory addresses (I haven't updated Helix in a while, though). I'll update it one of these days and see.

What about you? You're also seeing memory addresses?

**Comment by rkshthrmsh** [comment]
@weiying-chen and @David-Else, I tried it today and it seems like values for variables on the stack are shown but only addresses for the heap?
<img width="711" alt="image" src="https://github.com/helix-editor/helix/assets/16115390/67c4910e-964a-4c75-8c0c-8e2dcb459783">

**Comment by HealsCodes** [comment]
For those here that are trying to debug with helix on macOS:

- install [homebrew](https://brew.sh) if you haven't already, then run
- `brew install llvm && ln -s $(brew --prefix)/opt/llvm/bin/lldb-vscode $(brew --prefix)/bin/`

That'll get you a usable lldb-vscode in your homebrew path without messing up the system llvm / lldb

**Reply by rkshthrmsh** [reply]
Hi @Shirk, are you able to circumvent the issues mentioned above (regarding inability to see variable values) by installing lldb-vscode this way?

**Reply by HealsCodes** [reply]
No, it doesn't - it's just a quick and clean means to get lldb-vscode as I felt that wasn't really addressed yet but asked by the OP.
It still shows addresses for allocated locals:
<img width="817" alt="Screenshot 2023-07-29 at 14 25 54@2x" src="https://github.com/helix-editor/helix/assets/304352/f21a6d08-7c3d-4cca-87f5-51a8fcc2afe5">

(that's on helix 23.05 (7f5940be) / llvm 16.0.6)

**Reply by HealsCodes** [reply]
It looks like the output is down to lldb-vscode having no knowledge of the two commands required to setup rust-pretty printing.

```lldb
(lldb) command script import "/Users/USER/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/etc/lldb_lookup.py"
(lldb) command source -s 0 '/Users/USER/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/etc/lldb_commands'
```

^-- these two commands get run by `rust-lldb` as session setup and they enable the decoding of rust parameters into a readable version with actual values.

See [my comment](https://github.com/helix-editor/helix/issues/7007#issuecomment-1656758559) on #7007 for details.

**Reply by quantonganh** [reply]
@rkshthrmsh Please check the updated wiki: https://github.com/helix-editor/helix/wiki/Debugger-Configurations#rust

**Reply by rkshthrmsh** [reply]
@quantonganh, thanks a ton. It works pretty great now!
@David-Else and @weiying-chen, please note, in case you're still interested.

**Reply by David-Else** [reply]
@quantonganh  Thanks for all your awesome work on the debugger WIKI. I just had a go at re-writing the whole thing to make it more readable and easy to follow for users. I hope I didn't mess anything up! Please could you have a check check to see if you agree with any changes I made?

https://github.com/helix-editor/helix/wiki/Debugger-Configurations/35a284492c002d1d6d9dc923a8c83f586f03409d

or a direct diff:

https://github.com/helix-editor/helix/wiki/Debugger-Configurations/_compare/bf4d6161e14c2f37c7c68dd3334470ffb1e1ef13...35a284492c002d1d6d9dc923a8c83f586f03409d

Thanks!

**Comment by David-Else** [comment]
@rkshthrmsh Thanks!

**Comment by RandomExplosion** [comment]
It looks like lldb-vscode (included in `lldb` itself) is different from vscode-lldb (aka codelldb)
They seem to be two completely different debuggers, which should we be using?

**Reply by kirawi** [reply]
The default config uses `lldb-vscode`, but there is also a guide for `codelldb` on the wiki: https://github.com/helix-editor/helix/wiki/Debugger-Configurations#codelldb-
